### ISSUE TITLE
Unclear code in Mask Loss Function

### ISSUE THREAD

**Issue by divgill91** [hustvl/SparseInst#49]
in the loss function for mask generation there is this segment of code

In the loss.py:loss_masks_with_iou_objectness

`        # FIXME: tgt_idx
 mix_tgt_idx = torch.zeros_like(tgt_idx[1])
 cum_sum = 0
 for num_mask in num_masks:
    mix_tgt_idx[cum_sum: cum_sum + num_mask] = cum_sum
    cum_sum += num_mask
 mix_tgt_idx += tgt_idx[1]
`

I am not sure what it's purpose is. Such a thing is not doe for the classes head or the objectness head, just the masks. Why is that ?

Thanks

**Comment by wondervictor** [comment]
Hi @divgill91, thanks for your interest in SparseInst!
We need to calculate the pixel/dice loss between *matched* predicted masks and ground-truth masks. This code snippet stacks ground-truth masks (by index) from a batch of images into a KxHxW tensor (K is the sum of all masks from a batch). Each image contains different numbers of ground-truth masks and we need to gather all masks for loss computation. Directly stacking all masks with padding (into a BxNxHxW tensor) will incur large memory waste, which is adopted in DETR.

**Comment by wondervictor** [comment]
We believe we have answered your question, and as such I'm closing this issue, but let us know if you have further questions. Thanks for your interest in SparseInst 😊 . If you find it useful in your work or research, could you give us a star 🌟 or help us recommend SparseInst to your friends.
